### PR TITLE
Add hotunplug virtio scsi controller and add multi storage hotplug/unplug

### PR DESCRIPTION
**Pull request by peixiu**
1. Original code only have hotplug virtio scsi controller,
but have not hotunplug virtio scsi controller function.
It's a basic requirement of hotunplug cases.
2. And also update the code to match the requirement of the
case "hotplug/unplug multi storage".

id: 1520877, 1542350, 1231141

Signed-off-by: Peixiu Hou <phou@redhat.com>

### CODE REVIEW

**Comment by peixiu** [comment]
also tried test with virtio_blk system disk + block_scsi disk, all cases can be passed.

(01/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.one_pci: PASS (159.88 s)
 (02/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.multi_pci: PASS (373.01 s)
 (03/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_repetition.one_pci: PASS (223.20 s)
 (04/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_repetition.multi_pci: PASS (567.20 s)
 (05/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_reboot.one_pci: PASS (467.30 s)
 (06/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_reboot.multi_pci: PASS (649.19 s)
 (07/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_shutdown.after_plug.one_pci: PASS (81.93 s)
 (08/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_shutdown.after_plug.multi_pci: PASS (144.30 s)
 (09/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_shutdown.after_unplug.one_pci: PASS (93.24 s)
 (10/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_shutdown.after_unplug.multi_pci: PASS (178.63 s)
 (11/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_stop_continue.one_pci: PASS (193.73 s)
 (12/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_stop_continue.multi_pci: PASS (412.45 s)
 (13/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_system_reset.one_pci: PASS (470.67 s)
 (15/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.one_pci: PASS (104.05 s)
 (16/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.multi_pci: PASS (197.06 s)
 (17/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_repetition.one_pci: PASS (102.30 s)
 (18/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_repetition.multi_pci: PASS (192.82 s)
 (19/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_reboot.one_pci: PASS (195.20 s)
 (20/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_reboot.multi_pci: PASS (289.30 s)
 (21/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_shutdown.after_plug.one_pci: PASS (99.38 s)
 (22/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_shutdown.after_plug.multi_pci: PASS (176.87 s)
 (23/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_shutdown.after_unplug.one_pci: PASS (102.27 s)
 (24/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_shutdown.after_unplug.multi_pci: PASS (196.13 s)
 (25/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_stop_continue.one_pci: PASS (116.56 s)
 (26/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_stop_continue.multi_pci: PASS (212.03 s)
 (27/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_system_reset.one_pci: PASS (209.27 s)
 (28/28) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_system_reset.multi_pci: PASS (304.81 s)

**Comment by peixiu** [comment]
Tested on RHEL7.5, all cases can be passed.

 (01/30) smp_4.4096m.repeat1.pre_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads: PASS (1532.48 s)
 (02/30) smp_4.4096m.repeat1.pre_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.rh_kernel_update: PASS (362.65 s)
 (03/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.one_pci: PASS (96.56 s)
 (04/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.multi_pci: PASS (212.21 s)
 (05/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_repetition.one_pci: PASS (135.29 s)
 (06/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_repetition.multi_pci: PASS (330.68 s)
 (07/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_reboot.one_pci: PASS (291.10 s)
 (08/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_reboot.multi_pci: PASS (411.06 s)
 (09/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_shutdown.after_plug.one_pci: PASS (42.82 s)
 (10/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_shutdown.after_plug.multi_pci: PASS (62.05 s)
 (11/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_shutdown.after_unplug.one_pci: PASS (54.50 s)
 (12/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_shutdown.after_unplug.multi_pci: PASS (95.50 s)
 (13/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_stop_continue.one_pci: PASS (141.44 s)
 (14/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_stop_continue.multi_pci: PASS (259.31 s)
 (15/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_system_reset.one_pci: PASS (324.91 s)
 (16/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_system_reset.multi_pci: PASS (443.64 s)
 (17/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.one_pci: PASS (47.41 s)
 (18/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.multi_pci: PASS (73.87 s)
 (19/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_repetition.one_pci: PASS (47.82 s)
 (20/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_repetition.multi_pci: PASS (75.13 s)
 (21/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_reboot.one_pci: PASS (113.59 s)
 (22/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_reboot.multi_pci: PASS (137.16 s)
 (23/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_shutdown.after_plug.one_pci: PASS (38.29 s)
 (24/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_shutdown.after_plug.multi_pci: PASS (51.13 s)
 (25/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_shutdown.after_unplug.one_pci: PASS (48.92 s)
 (26/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_shutdown.after_unplug.multi_pci: PASS (73.29 s)
 (27/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_stop_continue.one_pci: PASS (62.75 s)
 (28/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_stop_continue.multi_pci: PASS (88.40 s)
 (29/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_system_reset.one_pci: PASS (124.42 s)
 (30/30) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_system_reset.multi_pci: PASS (149.30 s)

File: qemu/tests/cfg/block_hotplug.cfg [hunk 1]
```diff
@@ -26,14 +26,18 @@
...
             boot_drive_stg1 = no
+            boot_drive_stg2 = no
             image_name_stg1 = images/storage1
-            image_size_stg1 = 1G
+            image_name_stg2 = images/storage2
+            image_size_equal = 1G
             remove_image_stg1 = yes
+            remove_image_stg2 = yes
             force_create_image_stg1 = yes
+            force_create_image_stg2 = yes
```
Comment (xiagao) [inline qemu/tests/cfg/block_hotplug.cfg]: why not add stg2 in other variants,such as 
 - block_virtio:
- fmt_qcow2:

File: qemu/tests/block_hotplug.py [hunk 2]
```diff
@@ -90,6 +93,7 @@ def get_new_disk(disk1, disk2):
                 disks_before_plug = find_disk(vm, get_disk_cmd)
 
                 if params.get("need_controller", "no") == "yes":
+                    remove_controller = True
```
Comment (hereischen) [inline qemu/tests/block_hotplug.py]: just want to check with you, does `need_controller` mean the guest needs hot-plug controller ?

File: qemu/tests/cfg/block_hotplug.cfg [hunk 3]
```diff
@@ -26,14 +26,18 @@
             blk_num = 1
             repeat_times = 3
         - multi_pci:
-            blk_num = 2
+            blk_num = 3
             repeat_times = 3
-            images += " stg1"
+            images += " stg1 stg2"
             boot_drive_stg1 = no
+            boot_drive_stg2 = no
```
Comment (hereischen) [inline qemu/tests/cfg/block_hotplug.cfg]: Put these params into .py as optional params might be a good idea.

File: qemu/tests/block_hotplug.py (hunk 2 above)
Comment (peixiu) [reply qemu/tests/block_hotplug.py]: hachen, yes

File: qemu/tests/cfg/block_hotplug.cfg (hunk 1 above)
Comment (peixiu) [reply qemu/tests/cfg/block_hotplug.cfg]: Thanks for question, tested with qcow2 format with this code, the result is match currently requirement, but it seems not dissonance, will  optimization the code, thanks~

File: qemu/tests/cfg/block_hotplug.cfg (hunk 3 above)
Comment (peixiu) [reply qemu/tests/cfg/block_hotplug.cfg]: Hi hachen, 
it's a param based to the case multi_pci, if test with_plug, boot_drive_stg* = no, then hotplug stg0, stg1, stg2; if test without_plug, boot_drive_stg*=yes, it means will add disk at booting the guest up, then do hotunplug stg0, stg1, stg2. What do you mean that "put these params into .py as optional params"? if added to .py, if also need to add some condition to differ cases for one_pci and cases for multi_pci, if yes, if it's more  complex these case?

Thanks~

Comment (hereischen) [reply qemu/tests/cfg/block_hotplug.cfg]: OK,Let's keep it this way.

**Comment by peixiu** [comment]
Hi hachen, xiaogao, I updated the code make the code match the qcow2 and raw format, could you help to review it again? thanks~

smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.with_plug.with_reboot.multi_pci: PASS smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_qcow2.without_plug.with_reboot.multi_pci: PASS
smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_raw.with_plug.with_reboot.multi_pci: PASS smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.block_hotplug.block_scsi.fmt_raw.without_plug.with_reboot.multi_pci: PASS

**Comment by CongLi** [comment]
@peixiu 

Could you please rebase your commit?

Thanks.

**Comment by peixiu** [comment]
Due to part of code conflict with the patch https://github.com/autotest/tp-qemu/pull/1168, close this pull request and rebase the patch for multi storage hotplug/unplug part.

Thanks~

**Pull request closed without merging**